dashboard/app/assets/javascripts/levels/contract_match.js.jsx

```diff
@@ -0,0 +1,257 @@
+$(window).load(function () {
+
+  /**
+   * @param fn function to be curried with arguments
+   * @params [...] rest of args
+   * @returns {Function} new function with the given arguments pre-set as the
+   *                     defaults (curried)
+   */
+  var curry = function(fn) {
+    var args = Array.prototype.slice.call(arguments, 1);
+
+    return function () {
+      return fn.apply(this, args.concat(
+        Array.prototype.slice.call(arguments, 0)));
+    };
+  };
+
+  /**
+   * Enum of block types. Used for block and domain/range coloring
+   * @enum {string}
+   */
+  var blockValueType = {
+    NONE: 'None',
+    NUMBER: 'Number',
+    STRING: 'String',
+    IMAGE: 'Image',
+    BOOLEAN: 'Boolean'
+  };
+
+  var typesToColors = {};
+  typesToColors[blockValueType.NONE] = "#999999";
+  typesToColors[blockValueType.NUMBER] = "#00ccff";
+  typesToColors[blockValueType.STRING] = "#009999";
+  typesToColors[blockValueType.IMAGE] = "#9900cc";
+  typesToColors[blockValueType.BOOLEAN] = "#336600";
+
+  /**
+   * Component Structure:
+   *
+   * - ContractForm
+   *   - DomainsList
+   *     - TypeChooser
+   */
+  var ContractForm = React.createClass({
+    getName: function () {
+      return this.state.name;
+    },
+    getRangeType: function () {
+      return this.state.rangeType;
+    },
+    getDomainTypes: function () {
+      return this.state.domainTypes;
+    },
+    maxDomainID: 0,
+    getInitialState: function () {
+      /**
+       * @param data {{
+       *   name: string,
+       *   rangeType: BlockValueType,
+       *   domainTypes: {
+       *     type: BlockValueType,
+       *     key: string
+       *   },
+       * }}
+       */
+      return {
+        name: "",
+        rangeType: blockValueType.NONE,
+        domainTypes: [
+          {
+            key: 'domain' + (this.maxDomainID++),
+            type: blockValueType.NONE,
+            order: 0
+          }
+        ]
```

Comment:
Feels a little weird to be generating ids in to different places. Might also call it maxDomainID_ to make it clear it shouldnt be accessed elsewhere - or just make it a local.